07_VirtualFileSystem/03_TarFileSystem.md

```diff
@@ -0,0 +1,158 @@
+
+
+```c
+int octascii_to_dec(char *number, int size);
+```
+
+The size parameter tells us how many bytes is the digit long, and in the case of a tar object record the size is fixed: 12 bytes. Since we just need to implement a data structure for the header, this is left as exercise. Let's assume just that a new type is defined with the name `tar_record`.
+
+
+### Searching for a file
+
+Since the tar format doesn't have any file table or linked lists, or similar to search for files, we need everytime to start from the first record and scan one after each other, if the record is found we will return the pointer to it, otherwise we will eventually reach the end of the archive (file system in our case) meaning that the file searched is not present. 
+
+The picture below show how data is stored into a tar archive. 
+
+
+![Tar Archive](Images/tar.png)
+
+To move from the first header to the next we simply need to use the following formula:
+
+$$ next\_header = header\_ptr + header\_size + file\_size $$
+
+The lookup function then will be in the form of a loop.  But how to tell if we have reached the end of the archive? As mentioned above, if there are two or more zero-filled records, it indicated the end. So while searching, we need to make sure that we keep track of the number of zeroed records. The main lookup loop should be similar to the following pseudo-code:
+
+```c 
+int zero_counter = 0;
+while (zero_counter < 2) {
+    if (is_zeroed(current_record) ) {
+        zero_counter++;
+        continue;
+    }
+    zero_counter = 0;
+    //lookup for file or load next record if not found
+}
+```
+
+The `is_zeroed` function is a helper function that we should implement, as the name suggest it should just check that the current record is full of zeros, the implementation is left as an exercise. Within the loop now we just need to search for the file requested, the tricky part here is that we can have two scenarios:
+
+* The filename length is less than 100 bytes, in this case it is stored into the `file_name` field
+* The length is greater than 100 bytes (up to 256) so in this case the filename is split in two parts, the first 100 bytes goes into the `file_name` field, the rest goes into the `filename_prefix` field. 
+
+An easy solution is to check first the searched filename length, if it less than 100 characters, so we can use just the `file_name` field, otherwise we can merge the two fields and compare than with the searched filename. The updated loop pseudo-code should look similar to this: , 
+
+```c
+
+char tar_filename[256];
+int zero_counter = 0;
+//The starting address should be known somehow to the OS)
+tar_record current_record = tar_fs_start_address; 
+while (zero_counter < 2) {
+    if (is_zeroed(current_record) ) {
+        zero_counter++;
+        continue;
+    }
+    zero_counter = 0;
+    if ( tar_record.filename_prefix[0] != 0) {
```

Comment:
tar_record is the type! not variable name, you've used this a few times down below too.